legend-engine-xt-persistence-test-runner/src/main/java/org/finos/legend/engine/testable/persistence/mapper/DeriveDatasets.java

```diff
@@ -0,0 +1,176 @@
+                .group(mainDataset.datasetReference().group())
+                .alias(mainDataset.datasetReference().alias().orElse(null));
+
+        this.mainSchemaDefinitionBuilder = SchemaDefinition.builder()
+                .addAllFields(mainDataset.schema().fields())
+                .addAllIndexes(mainDataset.schema().indexes())
+                .shardSpecification(mainDataset.schema().shardSpecification())
+                .columnStoreSpecification(mainDataset.schema().columnStoreSpecification());
+
+        this.stagingSchemaDefinitionBuilder = SchemaDefinition.builder()
+                .addAllFields(mainDataset.schema().fields())
+                .addAllIndexes(mainDataset.schema().indexes())
+                .shardSpecification(mainDataset.schema().shardSpecification())
+                .columnStoreSpecification(mainDataset.schema().columnStoreSpecification());
+
+        this.stagingDatasetBuilder = JsonExternalDatasetReference.builder()
+                .name(mainDataset.datasetReference().name().get() + STAGING_SUFFIX)
+                .database(mainDataset.datasetReference().database())
+                .group(mainDataset.datasetReference().group())
+                .alias(mainDataset.datasetReference().alias().isPresent() ? mainDataset.datasetReference().alias().get() + STAGING_SUFFIX : null)
+                .data(testData);
+    }
+
+    @Override
+    public Datasets visit(AppendOnly appendOnly)
+    {
+        Dataset stagingDataset = stagingDatasetBuilder.schema(stagingSchemaDefinitionBuilder.build()).build();
+        if (appendOnly.filterDuplicates) mainSchemaDefinitionBuilder = enrichWithDigest();
+        mainSchemaDefinitionBuilder = appendOnly.auditing.accept(new MappingVisitors.EnrichSchemaWithAuditing(mainSchemaDefinitionBuilder, mainDataset));
+        Dataset enrichedMainDataset = mainDatasetDefinitionBuilder.schema(mainSchemaDefinitionBuilder.build()).build();
+
+        return Datasets.of(enrichedMainDataset, stagingDataset);
+    }
+
+    @Override
+    public Datasets visit(BitemporalDelta bitemporalDelta)
+    {
+        ValidityDerivation validityDerivation = bitemporalDelta.validityMilestoning.accept(MappingVisitors.MAP_TO_COMPONENT_VALIDITY_MILESTONING).validityDerivation();
+
+        stagingSchemaDefinitionBuilder = validityDerivation.accept(new MappingVisitors.EnrichSchemaWithValidityMilestoningDerivation(stagingSchemaDefinitionBuilder, mainDataset));
+        stagingSchemaDefinitionBuilder = bitemporalDelta.mergeStrategy.accept(new MappingVisitors.EnrichSchemaWithMergyStrategy(stagingSchemaDefinitionBuilder, mainDataset));
+        Dataset stagingDataset = stagingDatasetBuilder.schema(stagingSchemaDefinitionBuilder.build()).build();
+
+        mainSchemaDefinitionBuilder = enrichWithDigest();
+        mainSchemaDefinitionBuilder = validityDerivation.accept(new MappingVisitors.EnrichSchemaWithValidityMilestoningDerivation(mainSchemaDefinitionBuilder, mainDataset));
```

Comment:
Do we need to add sourceFrom and sourceThrough in main table ?

Reply:
for now yes - we will address this in the later ticket

Reply:
Understood